### DISCUSSION TITLE
Type resolution in generator module

### DISCUSSION THREAD

**Discussion by marschwar** [detekt/detekt#3628] [Q&A] [answered]
This question is related to #3608 . 

To avoid the duplication of default config values in rules I would like to extract the default value that is shown in the documentation from the delegated property in the `RuleVisitor`. Let me use an example.

```Kotlin
class RuleA(config: Config = Config.empty) : Rule(config) {
    // no problem without type resolution
    private val configA: Int by config(defaultValue = 99)

    // a little more complicated but still possible without type resolution
    private val configB: Int by config(defaultValue = DEFAULT_CONFIG_B)

    // ???
    private val configC: Int by config(defaultValue = DEFAULT_DEFINED_SOMEWHERE_ELSE)

    ...

    companion object {
        const val DEFAULT_CONFIG_B = 15
    }
}
```  

I have no idea how to extract the default value for the documentation if it is defined somewhere else in the code base without type resolution. A real life example is `CyclomaticComplexity.DEFAULT_NESTING_FUNCTIONS`

My  preference would be to avoid type resolution here. Any ideas? 
If I actually do need it, what would I have to do in order to create a `BindingContext`?

**Comment by cortinico** [comment, marked as answer]
> My preference would be to avoid type resolution here. Any ideas?

That's not possible, unless we restrict how default value are defined.
What I think it's further complicated to address here is that `defaultValue` could potentially be a function, so you actually need to execute code in order to compute the default value. I'm unsure we want to go this way.

>  If I actually do need it, what would I have to do in order to create a BindingContext?

You should populate the `classpath` parameter correctly. See https://detekt.github.io/detekt/type-resolution.html
